### PR TITLE
Support Object opts[,Function cb] signature

### PR DESCRIPTION
**Pull request by golyshevd**
Now, `got` supports also (opts[, cb]) signature. Url is not required.

Example:

```js
got({host: 'example.com', path: '/example'}, function (err, data, res) {
   // stuff
});
```

Convenient thing to keep (hostname, port, etc.) in project config separately, and no need to format `url`.

### CODE REVIEW

**Comment by floatdrop** [comment]
Let start from simple - split this PR: one for `query`, one for object as first argument.

**Comment by golyshevd** [comment]
Removed query feature commit

**Comment by floatdrop** [comment]
Okay, could you squash commits, provide clear description of PR (what it does, why you need it)?

**Comment by golyshevd** [comment]
Ok, I will patch readme tonight and then squash commits after review

File: readme.md [hunk 1]
```diff
@@ -39,11 +39,10 @@ fs.createReadStream('index.html').pipe(got.post('todomvc.com'));
 
 It's a `GET` request by default, but can be changed in `options`.
 
-#### got(url, [options], [callback])
+#### got([url][, options][, callback])
```
Comment (floatdrop) [inline readme.md]: So `got()` is valid now?

Comment (golyshevd) [inline readme.md]: I did not thought about it earlier but in fact, yes, it will be like `got({})`. Need check for "at least one argument"?

Comment (floatdrop) [inline readme.md]: > `got([url][, options][, callback])`

I don't think this is a good API.

**Comment by floatdrop** [comment]
I like idea of passing `url.parse` object as first argument - this will simplify usecase of wrapping got in other module, that already parsed url string. But this should now take so much changes to implement.

**Comment by golyshevd** [comment]
What problem that it is few additional lines of code?

Simplest way to do it is to convert (opts, cb) to ('', opts, cb) but it is ugly in my opinion, and not quite right

**Comment by floatdrop** [comment]
@golyshevd problem is that it is not a _few_ lines. Less code to maintain and debug is always better imo.

**Comment by golyshevd** [comment]
You are absolutely right. But "many code" does not mean "complex code". In my opinion `got` becomes to be complex because of `got` is too polymorphic. But refactoring solutions is not a topic.

Personally, I need this feature in `got` and I have a plans to use `got` in future. But I have no idea how to do it with less diff

**Pull request closed without merging**